File: internal/toc/toc.go
package toc

import (
	"fmt"
	"html/template"
	"strings"

	"github.com/yuin/goldmark/ast"
)

type Heading struct {
	Level int
	ID    string
	Text  string
}

func Extract(doc ast.Node, source []byte) []Heading {
	var headings []Heading

	ast.Walk(doc, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
		if !entering {
			return ast.WalkContinue, nil
		}

		h, ok := n.(*ast.Heading)
		if !ok {
			return ast.WalkContinue, nil
		}

		id := ""
		if idAttr, found := h.AttributeString("id"); found {
			id = string(idAttr.([]byte))
		}

		txt := extractText(h, source)

		headings = append(headings, Heading{
			Level: h.Level,
			ID:    id,
			Text:  txt,
		})

		return ast.WalkContinue, nil
	})

	return headings
}

func extractText(n ast.Node, source []byte) string {
	var buf strings.Builder
	ast.Walk(n, func(child ast.Node, entering bool) (ast.WalkStatus, error) {
		if !entering {
			return ast.WalkContinue, nil
		}
		if t, ok := child.(*ast.Text); ok {
			buf.Write(t.Segment.Value(source))
		}
		return ast.WalkContinue, nil
	})
	return buf.String()
}

func RenderTOC(headings []Heading) template.HTML {
	if len(headings) == 0 {
		return ""
	}

	var buf strings.Builder
	buf.WriteString(`<nav class="toc">`)
	buf.WriteString("\n<ul>\n")

	prevLevel := 0
	for _, h := range headings {
		if prevLevel == 0 {
			prevLevel = h.Level
		}

		if h.Level > prevLevel {
			for i := prevLevel; i < h.Level; i++ {
				buf.WriteString("<ul>\n")
			}
		} else if h.Level < prevLevel {
			for i := h.Level; i < prevLevel; i++ {
				buf.WriteString("</li>\n</ul>\n")
			}
			buf.WriteString("</li>\n")
		} else if prevLevel > 0 {
			buf.WriteString("</li>\n")
		}

		buf.WriteString(fmt.Sprintf(`<li><a href="#%s">%s</a>`, h.ID, h.Text))
		prevLevel = h.Level
	}

	for i := headings[0].Level; i <= prevLevel; i++ {
		buf.WriteString("</li>\n</ul>\n")
	}

	buf.WriteString("</nav>\n")
	return template.HTML(buf.String())
}

File: src/toc.ts
export interface TocHeading {
  level: number;
  id: string;
  text: string;
}

export function extractHeadings(html: string): TocHeading[] {
  const headings: TocHeading[] = [];
  const re = /<h([2-6])\s+id="([^"]*)"[^>]*>(.*?)<\/h\1>/gi;
  let match;

  while ((match = re.exec(html)) !== null) {
    headings.push({
      level: parseInt(match[1], 10),
      id: match[2],
      text: stripHtml(match[3]),
    });
  }

  return headings;
}

function stripHtml(s: string): string {
  return s.replace(/<[^>]+>/g, "");
}

export function renderTOC(headings: TocHeading[]): string {
  if (headings.length === 0) return "";

  let html = '<nav class="toc">\n';
  let prevLevel = 0;

  for (const h of headings) {
    if (prevLevel === 0) {
      html += "<ul>\n";
      prevLevel = h.level;
    }

    if (h.level > prevLevel) {
      for (let i = prevLevel; i < h.level; i++) {
        html += "<ul>\n";
      }
    } else if (h.level < prevLevel) {
      for (let i = h.level; i < prevLevel; i++) {
        html += "</li>\n</ul>\n";
      }
      html += "</li>\n";
    } else if (prevLevel > 0) {
      html += "</li>\n";
    }

    html += `<li><a href="#${h.id}">${h.text}</a>`;
    prevLevel = h.level;
  }

  for (let i = headings[0].level; i <= prevLevel; i++) {
    html += "</li>\n</ul>\n";
  }

  html += "</nav>\n";
  return html;
}
